### ISSUE TITLE
Add custom rules to depot

### ISSUE THREAD

**Issue by stalkerg** [game-ci/steam-deploy#67]
Depot file it's not just folder with a game, it's have much more important rules like `FileProperties` or `FileExclusion` how can I add a such options to generated depot?

**Comment by webbertakken** [comment]
Good shout!

You'd have to extend the action to support this kind of functionality. This repo is open for everyone to contribute to and makes for a stable central logic that everyone can use.

**Comment by GabLeRoux** [comment]
If I understand correctly, the question can be summarized like this:

> How can I customize the Github Actions parameters to add options like `FileProperties` or `FileExclusion` to the Steam depot using `steam-deploy` action to support game folders with special configurations?

I'd say it's important to have a good understanding of how github actions work first. Here's a great starting point: 
 
https://docs.github.com/en/actions/creating-actions

The steps required to customize and add `FileProperties` and `FileExclusion` inputs to the action should look like this:

1. [Fork the repository](https://github.com/game-ci/steam-deploy/fork)
2. Add new inputs to your fork's [`actions.yml`](https://github.com/game-ci/steam-deploy/blob/main/action.yml). Would probably look something like this (draft):

    ```yaml
    inputs:
      ...
      fileProperties:
        description: 'Additional file properties to apply.'
        required: false
      fileExclusion:
        description: 'Patterns of files to exclude.'
        required: false
    ```

3. Add desired logic to [`steam_deploy.sh`](https://github.com/game-ci/steam-deploy/blob/main/steam_deploy.sh). 

    <details>
     <summary>Here's an untested draft</summary>
    
    ```bash
    #!/bin/bash
    set -euo pipefail
    IFS=$'\n\t'
    
    steamdir=${STEAM_HOME:-$HOME/Steam}
    contentroot=$(pwd)/$rootPath
    
    mkdir BuildOutput
    manifest_path=$(pwd)/manifest.vdf
    
    echo ""
    echo "#################################"
    echo "#   Generating Depot Manifests  #"
    echo "#################################"
    echo ""
    
    if [ -n "$firstDepotIdOverride" ]; then
      firstDepotId=$firstDepotIdOverride
    else
      firstDepotId=$((appId + 1))
    fi
    
    i=1;
    export DEPOTS="\n  "
    until [ $i -gt 9 ]; do
      eval "currentDepotPath=\$depot${i}Path"
      if [ -n "$currentDepotPath" ]; then
        currentDepot=$((firstDepotId + i - 1))
    
        echo ""
        echo "Adding depot${currentDepot}.vdf ..."
        echo ""
        export DEPOTS="$DEPOTS  \"$currentDepot\" \"depot${currentDepot}.vdf\"\n  "
        
        echo "\"DepotBuildConfig\"" > "depot${currentDepot}.vdf"
        echo "{" >> "depot${currentDepot}.vdf"
        echo "  \"DepotID\" \"$currentDepot\"" >> "depot${currentDepot}.vdf"
        echo "  \"FileMapping\"" >> "depot${currentDepot}.vdf"
        echo "  {" >> "depot${currentDepot}.vdf"
        echo "    \"LocalPath\" \"./$currentDepotPath/*\"" >> "depot${currentDepot}.vdf"
        echo "    \"DepotPath\" \".\"" >> "depot${currentDepot}.vdf"
        echo "    \"recursive\" \"1\"" >> "depot${currentDepot}.vdf"
        echo "  }" >> "depot${currentDepot}.vdf"
    
        # Handling File Exclusions
        if [ -n "${fileExclusion:-}" ]; then
          for exclusion in $fileExclusion; do
            echo "  \"FileExclusion\" \"$exclusion\"" >> "depot${currentDepot}.vdf"
          done
        fi
    
        # Handling File Properties
        if [ -n "${fileProperties:-}" ]; then
          echo "  \"FileProperties\"" >> "depot${currentDepot}.vdf"
          echo "  {" >> "depot${currentDepot}.vdf"
          for prop in $fileProperties; do
            echo "    \"$prop\"" >> "depot${currentDepot}.vdf"
          done
          echo "  }" >> "depot${currentDepot}.vdf"
        fi
    
        echo "}" >> "depot${currentDepot}.vdf"
    
        cat depot${currentDepot}.vdf
        echo ""
      fi
    
      i=$((i+1))
    done
    
    # ... Rest of the script remains unchanged ...
    ```
    
    </details>

4. Push changes to your fork

5. To test these changes, update your unity project's workflow file to point to your `steam-deploy` fork. It should look something like `your_github_username/steam-deploy@master` (replace `master` with your branch or tag where you added the changes).

If you plan to contribute changes back to the repository with a pull-request, make sure you follow the [contributing guidelines](https://github.com/game-ci/steam-deploy/blob/main/CONTRIBUTING.md).

Let us know if you have questions 🙏

**Comment by stalkerg** [comment]
Thanks everyone, I have no issue making it in my fork, I just put it here to keep track. I will check @GabLeRoux solution and if it works and usefully I will come back with PR.

**Comment by frostebite** [comment]
Added: game-ci/steam-deploy's thin-wrapper migration (#92) now includes an `extraExclusions` input, wired through to game-ci/cli's `deploy steam --extraExclusions` flag — comma-separated glob patterns excluded from the depot in addition to the built-in defaults (`*.pdb`, `*.log`, `*.vdf`, Burst debug/backup folders):

```yaml
- uses: game-ci/steam-deploy@v3
  with:
    extraExclusions: '*.pdb,Docs/*,*.tmp'
    # ...
```

This covers `FileExclusion` rules. Full `FileProperties` support (per-file overrides beyond exclusion, e.g. custom install flags on individual files) isn't implemented — the depot VDF generator (`plugins/steam-deploy/src/vdf-generator.ts` in game-ci/cli) is a small, self-contained function if you want to open a follow-up PR there for that specific case. Merging once #92 lands.

**Comment by frostebite** [comment]
Correction to my last comment — I checked Valve's actual depot build docs (https://partner.steamgames.com/doc/sdk/uploading) rather than assuming, and `extraExclusions` only covers part of what you asked for:

- **`FileExclusion`** — done, via the new `extraExclusions` input.
- **`FileProperties`** (`userconfig`/`versionedconfig` flags, for files a user/game modifies locally that shouldn't be clobbered by an update) — **not implemented**. Nothing in game-ci/cli's depot VDF generator sets these.
- **Multiple `FileMapping` blocks** — Valve's depot format supports more than one per depot (different `LocalPath`/`DepotPath`/`recursive` combinations), but game-ci/cli's `generateDepotVdf` (`plugins/steam-deploy/src/vdf-generator.ts`) hardcodes exactly one (`LocalPath: "./*"`, `DepotPath: "."`). That's arguably the bigger gap for "much more important rules" than exclusions alone — it's what you'd need for anything more structured than "ship the whole folder minus some globs."

Leaving this open — `extraExclusions` is a real but partial fix. Supporting `FileProperties` and multiple `FileMapping` entries would need a small API redesign in `generateDepotVdf` (an array of mappings/property rules instead of the current single-mapping shape) plus new CLI flags and wrapper inputs to expose them. That's a real follow-up, not something I want to rush through half-verified — happy to pick it up as a proper next PR if you confirm this is still the itch you're trying to scratch.

**Comment by frostebite** [comment]
Filed a proper follow-up: game-ci/cli#225 adds `fileMappings`/`fileProperties` to `generateDepotVdf` plus `--fileMapping "localPath=depotPath"` and `--fileProperty "localPath=userconfig|versionedconfig"` flags on `deploy steam` (both repeatable, verified against Valve's actual docs). Combined with `extraExclusions` (already merged), that covers `FileMapping`, `FileExclusion`, and `FileProperties` — the full set of depot customization Valve documents.

Once #225 merges, wiring these into this action's `action.yml`/inputs is the last step (mirrors how `extraExclusions` was wired here) — will follow up on that once the CLI side lands.
